include/ibamr/LevelSetUtilities.h

```diff
@@ -0,0 +1,435 @@
+     * would be the new context.
+     */
+    void fixMassLoss2PhaseFlows(double current_time,
+                                double new_time,
+                                bool skip_synchronize_new_state_data,
+                                int num_cycles,
+                                void* ctx);
+
+    /*!
+     * \brief Compute the value of the Lagrange multiplier \f$ q \f$ and use that to adjust the flow level set \f$
+     * \tilde{\phi \f$ when there is a solid phase in the domain with level set \f$ \psi < 0 \f$. Satisfies the
+     * constraint: \f$ f(q) = \int_{\Omega} H(\tilde{\phi} + q) H(\psi) \text{d}\Omega - V^0 = 0 \f$. where, \f$
+     * \tilde{\phi} \f$ is the level set field obtained from the reinitialization procedure, \f$ V^0 \f$ is the volume
+     * of the fluid at \f$ t = 0 \f$ s which needs to be conserved. Here, \f$ q \f$ is computed using the Newton's
+     * method till required tolerance. In practice the level set mass loss would be fixed during the postprocess
+     * integrate hierarchy stage. Hence the application time would be the new time and the variable context would be the
+     * new context.
+     */
+    void fixMassLoss3PhaseFlows(double current_time,
+                                double new_time,
+                                bool skip_synchronize_new_state_data,
+                                int num_cycles,
+                                void* ctx);
+
+    /*!
+     * \return Integrals of \f$ 1- H(\phi)\f$, \f$ H(\phi)\f$, and \f$ \delta(\phi) \f$  over
+     * the entire domain.
+     *
+     * Here, \f$ H(\phi) \f$ is the Heaviside function demarcating liquid and gas domains.
+     *
+     * \f$ \phi \f$ is taken to be positive in the liquid domain and negative in the gas domain.
+     *
+     * Physically, these three integrals represent volume of the gas region, liquid region, and
+     * surface area of the interface, respectively.
+     */
+    std::vector<double> computeHeavisideIntegrals2PhaseFlows(const LevelSetContainer lsc);
+
+    /*!
+     * \return Integrals of \f$ [1- H(\phi)] H(\Psi) \f$, \f$ H(\phi) H(\Psi)\f$, \f$ 1 - H(\Psi)\f$,
+     * and \f$ \delta(\phi)H(\Psi)\f$ over the entire domain.
+     *
+     * Here, \f$ H(\phi) \f$ is the Heaviside function demarcating liquid and gas domains,  and \f$ H(\Psi) \f$ is
+     * the Heaviside function demarcating solid and fluid (fluid = liquid and gas) domains.
+     *
+     * Physically, these four integrals represents volume of the gas region, liquid region, solid region, and
+     * surface area of the fluid interface, respectively.
+     *
+     * \f$ \phi \f$ is taken to be positive in the liquid domain and negative in the gas domain.
+     *
+     * \f$ \Psi \f$ is taken to be positive outside the solid and negative inside the solid.
+     */
+    std::vector<double> computeHeavisideIntegrals3PhaseFlows(const LevelSetContainer lsc);
```

Comment:
?
```suggestion
    std::vector<double> computeHeavisideIntegrals3PhaseFlows(const LevelSetContainer& lsc);
```

Reply:
Now ```d_ls_container``` is an object and not a pointer.

Reply:
Yes. You should take `lsc` by reference here instead of copying it.